docs/content/sips/017-configuring-and-emitting-observability.md

```diff
@@ -0,0 +1,85 @@
+title = "SIP 017 - Configuring and Emitting Observability"
+template = "main"
+date = "2024-02-27T12:00:00Z"
+
+---
+
+Summary: How to configure and emit telemetry to improve the runtime and trigger observability of Spin.
+
+Owner: caleb.schoepp@fermyon.com
+
+Created: February 27, 2024
+
+Updated: February 27, 2024
+
+## Background
+
+[Observability](https://opentelemetry.io/docs/concepts/observability-primer/#what-is-observability) is critical for a great developer experience. Improving the observability of Spin can be broken down into four categories:
+
+1. Runtime observability: Observing the Spin runtime itself e.g. spans and metrics around important parts of the Spin runtime.
+2. Trigger observability: Observing the requests made to Spin applications e.g. spans and metrics around the requests made to Spin applications.
+3. Component observability: Observing the interaction between composed components e.g having Wasmtime auto-instrument Wasm component graphs to produce spans.
+4. Guest observability: Observing code within the guest module e.g. building a host component for WASI Observe.
+
+More detail on these categories can be found [here](https://github.com/fermyon/spin/issues/2293).
+
+## Proposal
+
+This SIP aims to improve the runtime and trigger observability of Spin. First, it will suggest how observability in Spin can be configured. Second, it will present the specific runtime and trigger observability we want to emit. Finally, it will outline a plan going forward for how observability should be added to new Spin features.
+
+This proposal assumes that all observability data produced should be OTEL compliant.
+
+### Configuring observability
+
+The developer must be able to configure the endpoints of OTLP compliant collectors where traces and metrics can be sent. This can be expressed in the `spin.toml` manifest.
+
+```toml
+[application]
+otlp_grpc_traces_endpoint = "http://localhost:4317"
+otlp_grpc_metrics_endpoint = "http://localhost:4317"
+```
+
+**INPUT NEEDED:** How do we want to configure these endpoints? Where do they belong in the manifest? Do we want to allow for separate endpoints for traces and metrics? Do we want to allow for choosing between gRPC and HTTP? Should we let them pass a flag to `spin` to override the manifest value?
```

Comment:
This probably belongs in runtime config, not the app manifest.

Reply:
> Do we want to allow for choosing between gRPC and HTTP?

Probably yes, and we might want to actually start with `http/protobuf`: https://opentelemetry.io/docs/specs/otel/protocol/exporter/#specify-protocol

Reply:
> Probably yes, and we might want to actually start with / default to http/protobuf

Sounds fine. Why do you want to default to `http/protobuf` rather than gRPC?

Reply:
From the link:

> If no configuration is provided the default transport SHOULD be http/protobuf unless SDKs have good reasons to choose grpc as the default (e.g. for backward compatibility reasons when grpc was already the default in a stable SDK release).